### PR TITLE
HW-DAO-3

### PR DESCRIPTION
**Pull request by tbondare**

### CODE REVIEW

File: src/main/java/taxiapp/model/Car.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+public class Car {
+    private static Long lastAddedId;
+    private Long id;
+    private String model;
+    private Manufacturer manufacturer;
+
+    Car(String madel, Manufacturer manufacturer) {
+        this.id = lastAddedId == null ? 1 : ++lastAddedId;
+        this.model = madel;
```
Comment (YuraBondarenko) [inline src/main/java/taxiapp/model/Car.java:11]:
```suggestion
    Car(String model, Manufacturer manufacturer) {
        this.id = lastAddedId == null ? 1 : ++lastAddedId;
        this.model = model;
```

File: src/main/java/taxiapp/model/Car.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package taxiapp.model;
+
+public class Car {
+    private static Long lastAddedId;
+    private Long id;
+    private String model;
+    private Manufacturer manufacturer;
+
+    Car(String madel, Manufacturer manufacturer) {
+        this.id = lastAddedId == null ? 1 : ++lastAddedId;
```
Comment (YuraBondarenko) [inline src/main/java/taxiapp/model/Car.java:10]: ты можешь это не добавлять, так как ты сеттишь id в storage, когда создаешь новый manufacturer
так же и в остальных моделях

File: src/main/java/taxiapp/model/Driver.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package taxiapp.model;
+
+public class Driver {
+    private static Long lastAddedId;
+    private Long id;
+    private String name;
+    private String licenseNumber;
+
+    Driver(String name, String licenseNumber) {
+        this.id = lastAddedId == null ? 1 : ++lastAddedId;
```
Comment (YuraBondarenko) [inline src/main/java/taxiapp/model/Driver.java:10]: -

File: src/main/java/taxiapp/model/Manufacturer.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
+package taxiapp.model;
+
+public class Manufacturer {
+    private static Long lastAddedId;
+    private Long id;
+    private String name;
+    private String country;
+
+    public Manufacturer(String name, String country) {
+        this.id = lastAddedId == null ? 1 : ++lastAddedId;
```
Comment (YuraBondarenko) [inline src/main/java/taxiapp/model/Manufacturer.java:10]: -

File: src/main/java/taxiapp/dao/ManufacturerDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    public List<Manufacturer> getAllManufacturer() {
+        return Storage.manufacturers;
+    }
+
+    @Override
+    public Manufacturer update(Manufacturer manufacturer) {
+        Manufacturer oldValue = get(manufacturer.getId()).orElseThrow();
+        Manufacturer memberManufacture = new Manufacturer(manufacturer);
+        Storage.manufacturers.remove(oldValue);
+        Storage.manufacturers.add(memberManufacture);
```
Comment (YuraBondarenko) [inline src/main/java/taxiapp/dao/ManufacturerDaoImpl.java:36]:
```suggestion
        Storage.manufacturers.remove(oldValue);
        Storage.manufacturers.add(manufacturer);
```

**Review by YuraBondarenko — CHANGES_REQUESTED** [review]

**Review by YuraBondarenko — APPROVED** [review]

File: pom.xml [hunk 6]
```diff
@@ -9,8 +9,59 @@
...
+                    <version>3.8.0</version>
+                    <configuration>
+                        <source>${jdk.version}</source>
+                        <target>${jdk.version}</target>
+                        <encoding>${project.build.sourceEncoding}</encoding>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+    </build>
```
Comment (RomanDubovskyi) [inline pom.xml:66]: empty line is missing at the end of the file

File: src/main/java/taxiapp/dao/ManufacturerDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import taxiapp.lib.Dao;
+import taxiapp.model.Manufacturer;
+
+@Dao
+public class ManufacturerDaoImpl implements ManufacturerDao {
+
+    @Override
+    public Manufacturer create(Manufacturer manufacturer) {
+        Storage.addManufacturer(manufacturer);
+        return Storage.manufacturers.get(Storage.manufacturers.size() - 1);
```
Comment (RomanDubovskyi) [inline src/main/java/taxiapp/dao/ManufacturerDaoImpl.java:16]:
```suggestion
        return manufacturer;
```
it should be with id already at this point

File: src/main/java/taxiapp/dao/ManufacturerDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Override
+    public List<Manufacturer> getAllManufacturer() {
+        return Storage.manufacturers;
+    }
+
+    @Override
+    public Manufacturer update(Manufacturer manufacturer) {
+        Manufacturer oldValue = get(manufacturer.getId()).orElseThrow();
```
Comment (RomanDubovskyi) [inline src/main/java/taxiapp/dao/ManufacturerDaoImpl.java:33]:
```suggestion
        Manufacturer oldValue = get(manufacturer.getId()).get();
```
you can simply use get() if you don't want to throw a specific exception here

File: src/main/java/taxiapp/dao/ManufacturerDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        Storage.manufacturers.remove(oldValue);
+        Storage.manufacturers.add(manufacturer);
+        return oldValue;
+    }
+
+    @Override
+    public boolean delete(Long manufacturerId) {
+        return Storage.manufacturers.remove(get(manufacturerId).orElseThrow());
```
Comment (RomanDubovskyi) [inline src/main/java/taxiapp/dao/ManufacturerDaoImpl.java:41]: use removeIf method we want to return false if deletion has not been completed, not throw an exception

File: src/main/java/taxiapp/model/Car.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
+package taxiapp.model;
+
+public class Car {
+    private Long id;
+    private String model;
+    private Manufacturer manufacturer;
+
```
Comment (RomanDubovskyi) [inline src/main/java/taxiapp/model/Car.java:7]: car should have a list of drivers

File: src/main/java/taxiapp/service/ManufacturerService.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+import java.util.List;
+import java.util.Optional;
+import taxiapp.model.Manufacturer;
+
+public interface ManufacturerService {
+
+    Manufacturer create(Manufacturer manufacturer);
+
+    Optional<Manufacturer> get(Long manufacturerId);
```
Comment (RomanDubovskyi) [inline src/main/java/taxiapp/service/ManufacturerService.java:11]:
```suggestion
   Manufacturer get(Long manufacturerId);
```
on the service layer, we return an object not its optional

**Review by RomanDubovskyi — APPROVED** [review]
fix mistakes before merge

**Pull request merged**